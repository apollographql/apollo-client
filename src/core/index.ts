/* Core */

export {
  ApolloClient,
  ApolloClientOptions,
  DefaultOptions,
  mergeOptions,
} from './ApolloClient';
export {
  ObservableQuery,
  FetchMoreOptions,
  UpdateQueryOptions,
} from './ObservableQuery';
export {
  QueryOptions,
  WatchQueryOptions,
  MutationOptions,
  SubscriptionOptions,
  FetchPolicy,
  WatchQueryFetchPolicy,
  ErrorPolicy,
  FetchMoreQueryOptions,
  SubscribeToMoreOptions,
  MutationUpdaterFn,
} from './watchQueryOptions';
export { NetworkStatus } from './networkStatus';
export * from './types';
export {
  Resolver,
  FragmentMatcher,
} from './LocalState';
export { isApolloError, ApolloError } from '../errors';

/* Cache */

export {
  // All the exports (types and values) from ../cache, minus cacheSlot,
  // which we want to keep semi-private.
  Cache,
  ApolloCache,
  Transaction,
  DataProxy,
  InMemoryCache,
  InMemoryCacheConfig,
  MissingFieldError,
  defaultDataIdFromObject,
  ReactiveVar,
  makeVar,
  TypePolicies,
  TypePolicy,
  FieldPolicy,
  FieldReadFunction,
  FieldMergeFunction,
  FieldFunctionOptions,
  PossibleTypesMap,
} from '../cache';

export * from '../cache/inmemory/types';

/* Link */

export * from '../link/core';
export * from '../link/http';
export {
  fromError,
  toPromise,
  fromPromise,
  ServerError,
  throwServerError,
} from '../link/utils';

/* Utilities */

export {
  Observable,
  Observer,
  ObservableSubscription,
  Reference,
  isReference,
  makeReference,
  StoreObject,
} from '../utilities';

/* Supporting */

// The verbosity of invariant.{log,warn,error} can be controlled globally
// (for anyone using the same ts-invariant package) by passing "log",
// "warn", "error", or "silent" to setVerbosity. By default, Apollo Client
// displays warnings and errors, but hides invariant.log statements. Note
// that all invariant.* logging is hidden in production.
import { setVerbosity } from "ts-invariant";
export { setVerbosity as setLogVerbosity }
setVerbosity("warn");

// Note that importing `gql` by itself, then destructuring
// additional properties separately before exporting, is intentional.
// Due to the way the `graphql-tag` library is setup, certain bundlers
// can't find the properties added to the exported `gql` function without
// additional guidance (e.g. Rollup - see
// https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module).
// Instead of having people that are using bundlers with `@apollo/client` add
// extra bundler config to help `graphql-tag` exports be found (which would be
// awkward since they aren't importing `graphql-tag` themselves), this
// workaround of pulling the extra properties off the `gql` function,
// then re-exporting them separately, helps keeps bundlers happy without any
// additional config changes.
import gql from 'graphql-tag';
export const {
  resetCaches,
  disableFragmentWarnings,
  enableExperimentalFragmentVariables,
  disableExperimentalFragmentVariables
} = gql;
export { gql };
