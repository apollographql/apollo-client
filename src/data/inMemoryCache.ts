import { Cache, CacheWrite } from './cache';

import { DocumentNode } from 'graphql';

import { NormalizedCache } from './storeUtils';

import { ApolloReducerConfig } from '../store';

import { writeResultToStore } from './writeToStore';

import {
  readQueryFromStore,
  diffQueryAgainstStore,
  DiffResult,
} from './readFromStore';

export type OptimisticStoreItem = {
  id: string;
  data: NormalizedCache;
  transaction: (c: Cache) => void;
};

export class InMemoryCache extends Cache {
  private data: NormalizedCache;
  private config: ApolloReducerConfig;
  private optimistic: OptimisticStoreItem[] = [];

  constructor(config: ApolloReducerConfig, initialStore: NormalizedCache = {}) {
    super(config.addTypename || false);
    this.config = config;
    this.data = initialStore;
  }

  public getData(): NormalizedCache {
    return this.data;
  }

  public getOptimisticData(): NormalizedCache {
    if (this.optimistic.length === 0) {
      return this.data;
    }

    const patches = this.optimistic.map(opt => opt.data);
    return Object.assign({}, this.data, ...patches) as NormalizedCache;
  }

  public getOptimisticQueue(): OptimisticStoreItem[] {
    return this.optimistic;
  }

  public reset(): Promise<void> {
    this.data = {};

    return Promise.resolve();
  }

  public applyTransformer(
    transform: (i: NormalizedCache) => NormalizedCache,
  ): void {
    this.data = transform(this.data);
  }

  public diffQuery(query: {
    query: DocumentNode;
    variables: any;
    returnPartialData?: boolean;
    previousResult?: any;
    optimistic: boolean;
  }): DiffResult {
    return diffQueryAgainstStore({
      store: query.optimistic ? this.getOptimisticData() : this.data,
      query: query.query,
      variables: query.variables,
      returnPartialData: query.returnPartialData,
      previousResult: query.previousResult,
      fragmentMatcherFunction: this.config.fragmentMatcher,
      config: this.config,
    });
  }

  public read(query: {
    query: DocumentNode;
    variables: any;
    rootId?: string;
    previousResult?: any;
    optimistic: boolean;
  }): any {
    if (query.rootId && typeof this.data[query.rootId] === 'undefined') {
      return null;
    }

    return readQueryFromStore({
      store: query.optimistic ? this.getOptimisticData() : this.data,
      query: query.query,
      variables: query.variables,
      rootId: query.rootId,
      fragmentMatcherFunction: this.config.fragmentMatcher,
      previousResult: query.previousResult,
      config: this.config,
    });
  }

  public writeResult(write: CacheWrite): void {
    writeResultToStore({
      ...write,
      store: this.data,
      dataIdFromObject: this.config.dataIdFromObject,
      fragmentMatcherFunction: this.config.fragmentMatcher,
    });
  }

  public removeOptimistic(id: string) {
    // Throw away optimistic changes of that particular mutation
    const toPerform = this.optimistic.filter(item => item.id !== id);

    this.optimistic = [];

    // Re-run all of our optimistic data actions on top of one another.
    toPerform.forEach(change => {
      this.recordOptimisticTransaction(change.transaction, change.id);
    });
  }

  public performTransaction(transaction: (c: Cache) => void) {
    // TODO: does this need to be different, or is this okay for an in-memory cache?
    transaction(this);
  }

  public recordOptimisticTransaction(
    transaction: (c: Cache) => void,
    id: string,
  ) {
    const before = this.getOptimisticData();

    const orig = this.data;
    this.data = { ...before };
    transaction(this);
    const after = this.data;
    this.data = orig;

    const patch: any = {};

    Object.keys(after).forEach(key => {
      if (after[key] !== before[key]) {
        patch[key] = after[key];
      }
    });

    this.optimistic.push({
      id,
      transaction,
      data: patch,
    });
  }
}
