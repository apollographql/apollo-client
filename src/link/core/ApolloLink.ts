import { InvariantError, invariant } from 'ts-invariant';

import { Observable } from '../../utilities';
import {
  NextLink,
  Operation,
  RequestHandler,
  FetchResult,
  GraphQLRequest
} from './types';
import {
  validateOperation,
  createOperation,
  transformOperation,
} from '../utils';

function passthrough(op: Operation, forward: NextLink) {
  return (forward ? forward(op) : Observable.of()) as Observable<FetchResult>;
}

function toLink(handler: RequestHandler | ApolloLink) {
  return typeof handler === 'function' ? new ApolloLink(handler) : handler;
}

function isTerminating(link: ApolloLink): boolean {
  return link.request.length <= 1;
}

class LinkError extends Error {
  public link?: ApolloLink;
  constructor(message?: string, link?: ApolloLink) {
    super(message);
    this.link = link;
  }
}

export class ApolloLink {
  public static empty(): ApolloLink {
    return new ApolloLink(() => Observable.of());
  }

  public static from(links: (ApolloLink | RequestHandler)[]): ApolloLink {
    if (links.length === 0) return ApolloLink.empty();
    return links.map(toLink).reduce((x, y) => x.concat(y)) as ApolloLink;
  }

  public static split(
    test: (op: Operation) => boolean,
    left: ApolloLink | RequestHandler,
    right?: ApolloLink | RequestHandler,
  ): ApolloLink {
    const leftLink = toLink(left);
    const rightLink = toLink(right || new ApolloLink(passthrough));

    if (isTerminating(leftLink) && isTerminating(rightLink)) {
      return new ApolloLink(operation => {
        return test(operation)
          ? leftLink.request(operation) || Observable.of()
          : rightLink.request(operation) || Observable.of();
      });
    } else {
      return new ApolloLink((operation, forward) => {
        return test(operation)
          ? leftLink.request(operation, forward) || Observable.of()
          : rightLink.request(operation, forward) || Observable.of();
      });
    }
  }

  public static execute(
    link: ApolloLink,
    operation: GraphQLRequest,
  ): Observable<FetchResult> {
    return (
      link.request(
        createOperation(
          operation.context,
          transformOperation(validateOperation(operation)),
        ),
      ) || Observable.of()
    );
  }

  public static concat(
    first: ApolloLink | RequestHandler,
    second: ApolloLink | RequestHandler,
  ) {
    const firstLink = toLink(first);
    if (isTerminating(firstLink)) {
      invariant.warn(
        new LinkError(
          `You are calling concat on a terminating link, which will have no effect`,
          firstLink,
        ),
      );
      return firstLink;
    }
    const nextLink = toLink(second);

    if (isTerminating(nextLink)) {
      return new ApolloLink(
        operation =>
          firstLink.request(
            operation,
            op => nextLink.request(op) || Observable.of(),
          ) || Observable.of(),
      );
    } else {
      return new ApolloLink((operation, forward) => {
        return (
          firstLink.request(operation, op => {
            return nextLink.request(op, forward) || Observable.of();
          }) || Observable.of()
        );
      });
    }
  }

  constructor(request?: RequestHandler) {
    if (request) this.request = request;
  }

  public split(
    test: (op: Operation) => boolean,
    left: ApolloLink | RequestHandler,
    right?: ApolloLink | RequestHandler,
  ): ApolloLink {
    return this.concat(
      ApolloLink.split(test, left, right || new ApolloLink(passthrough))
    );
  }

  public concat(next: ApolloLink | RequestHandler): ApolloLink {
    return ApolloLink.concat(this, next);
  }

  public request(
    operation: Operation,
    forward?: NextLink,
  ): Observable<FetchResult> | null {
    throw new InvariantError('request is not implemented');
  }

  protected onError(
    error: any,
    observer?: ZenObservable.Observer<FetchResult>,
  ): false | void {
    if (observer && observer.error) {
      observer.error(error);
      // Returning false indicates that observer.error does not need to be
      // called again, since it was already called (on the previous line).
      // Calling observer.error again would not cause any real problems,
      // since only the first call matters, but custom onError functions
      // might have other reasons for wanting to prevent the default
      // behavior by returning false.
      return false;
    }
    // Throw errors will be passed to observer.error.
    throw error;
  }

  public setOnError(fn: ApolloLink["onError"]): this {
    this.onError = fn;
    return this;
  }
}
