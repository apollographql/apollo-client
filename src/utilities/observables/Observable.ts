export {
  Observable,
  PartialObserver as Observer,
  Subscription as ObservableSubscription,
} from "rxjs";
