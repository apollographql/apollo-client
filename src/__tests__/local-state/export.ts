import gql from 'graphql-tag';
import { print } from 'graphql/language/printer';

import { Observable } from '../../utilities/observables/Observable';
import { itAsync } from '../../utilities/testing/itAsync';
import { ApolloLink } from '../../link/core/ApolloLink';
import { ApolloClient } from '../..';
import { InMemoryCache } from '../../cache/inmemory/inMemoryCache';

describe('@client @export tests', () => {
  it(
    'should not break @client only queries when the @export directive is ' +
      'used',
    done => {
      const query = gql`
        {
          field @client @export(as: "someVar")
        }
      `;

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link: ApolloLink.empty(),
        resolvers: {},
      });

      cache.writeQuery({
        query,
        data: { field: 1 },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({ field: 1 });
        done();
      });
    },
  );

  it(
    'should not break @client only queries when the @export directive is ' +
      'used on nested fields',
    done => {
      const query = gql`
        {
          car @client {
            engine {
              torque @export(as: "torque")
            }
          }
        }
      `;

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link: ApolloLink.empty(),
        resolvers: {},
      });

      cache.writeQuery({
        query,
        data: {
          car: {
            engine: {
              cylinders: 8,
              torque: 7200,
              __typename: 'Engine',
            },
            __typename: 'Car',
          },
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          car: {
            engine: {
              torque: 7200,
            },
          },
        });
        done();
      });
    },
  );

  it(
    'should store the @client field value in the specified @export ' +
      'variable, and make it avilable to a subsequent resolver',
    done => {
      const query = gql`
        query currentAuthorPostCount($authorId: Int!) {
          currentAuthorId @client @export(as: "authorId")
          postCount(authorId: $authorId) @client
        }
      `;

      const testAuthorId = 100;
      const testPostCount = 200;

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        resolvers: {
          Query: {
            postCount(_, { authorId }) {
              return authorId === testAuthorId ? testPostCount : 0;
            },
          },
        },
      });

      cache.writeQuery({
        query,
        data: {
          currentAuthorId: testAuthorId,
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          currentAuthorId: testAuthorId,
          postCount: testPostCount,
        });
        done();
      });
    },
  );

  it(
    'should store the @client nested field value in the specified @export ' +
      'variable, and make it avilable to a subsequent resolver',
    done => {
      const query = gql`
        query currentAuthorPostCount($authorId: Int!) {
          currentAuthor @client {
            name
            authorId @export(as: "authorId")
          }
          postCount(authorId: $authorId) @client
        }
      `;

      const testAuthor = {
        name: 'John Smith',
        authorId: 100,
        __typename: 'Author',
      };

      const testPostCount = 200;

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        resolvers: {
          Query: {
            postCount(_, { authorId }) {
              return authorId === testAuthor.authorId ? testPostCount : 0;
            },
          },
        },
      });

      cache.writeQuery({
        query,
        data: {
          currentAuthor: testAuthor,
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          currentAuthor: testAuthor,
          postCount: testPostCount,
        });
        done();
      });
    },
  );

  it('should allow @client @export variables to be used with remote queries', done => {
    const query = gql`
      query currentAuthorPostCount($authorId: Int!) {
        currentAuthor @client {
          name
          authorId @export(as: "authorId")
        }
        postCount(authorId: $authorId)
      }
    `;

    const testAuthor = {
      name: 'John Smith',
      authorId: 100,
      __typename: 'Author',
    };

    const testPostCount = 200;

    const link = new ApolloLink(() =>
      Observable.of({
        data: {
          postCount: testPostCount,
        },
      }),
    );

    const cache = new InMemoryCache();
    const client = new ApolloClient({
      cache,
      link,
      resolvers: {},
    });

    cache.writeQuery({
      query,
      data: {
        currentAuthor: testAuthor,
      },
    });

    return client.query({ query }).then(({ data }: any) => {
      expect({ ...data }).toMatchObject({
        currentAuthor: testAuthor,
        postCount: testPostCount,
      });
      done();
    });
  });

  it(
    'should support @client @export variables that are nested multiple ' +
      'levels deep',
    done => {
      const query = gql`
        query currentAuthorPostCount($authorId: Int!) {
          appContainer @client {
            systemDetails {
              currentAuthor {
                name
                authorId @export(as: "authorId")
              }
            }
          }
          postCount(authorId: $authorId)
        }
      `;

      const appContainer = {
        systemDetails: {
          currentAuthor: {
            name: 'John Smith',
            authorId: 100,
            __typename: 'Author',
          },
          __typename: 'SystemDetails',
        },
        __typename: 'AppContainer',
      };

      const testPostCount = 200;

      const link = new ApolloLink(() =>
        Observable.of({
          data: {
            postCount: testPostCount,
          },
        }),
      );

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      cache.writeQuery({
        query,
        data: {
          appContainer,
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          appContainer,
          postCount: testPostCount,
        });
        done();
      });
    },
  );

  it('should ignore @export directives if not used with @client', done => {
    const query = gql`
      query currentAuthorPostCount($authorId: Int!) {
        currentAuthor {
          name
          authorId @export(as: "authorId")
        }
        postCount(authorId: $authorId)
      }
    `;

    const testAuthor = {
      name: 'John Smith',
      authorId: 100,
      __typename: 'Author',
    };
    const testPostCount = 200;

    const link = new ApolloLink(() =>
      Observable.of({
        data: {
          currentAuthor: testAuthor,
          postCount: testPostCount,
        },
      }),
    );

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link,
      resolvers: {},
    });

    return client.query({ query }).then(({ data }: any) => {
      expect({ ...data }).toMatchObject({
        currentAuthor: testAuthor,
        postCount: testPostCount,
      });
      done();
    });
  });

  itAsync(
    'should support setting an @client @export variable, loaded from the ' +
      'cache, on a virtual field that is combined into a remote query.',
    (resolve, reject) => {
      const query = gql`
        query postRequiringReview($reviewerId: Int!) {
          postRequiringReview {
            id
            title
            loggedInReviewerId @client @export(as: "reviewerId")
          }
          reviewerDetails(reviewerId: $reviewerId) {
            name
          }
        }
      `;

      const postRequiringReview = {
        id: 10,
        title: 'The Local State Conundrum',
        __typename: 'Post',
      };
      const reviewerDetails = {
        name: 'John Smith',
        __typename: 'Reviewer',
      };
      const loggedInReviewerId = 100;

      const link = new ApolloLink(({ variables }) => {
        expect(variables).toMatchObject({ reviewerId: loggedInReviewerId });
        return Observable.of({
          data: {
            postRequiringReview,
            reviewerDetails,
          },
        });
      }).setOnError(reject);

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      cache.writeQuery({
        query,
        data: {
          postRequiringReview: {
            loggedInReviewerId,
            __typename: 'Post',
            id: 10,
          },
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          postRequiringReview: {
            id: postRequiringReview.id,
            title: postRequiringReview.title,
            loggedInReviewerId,
          },
          reviewerDetails,
        });
      }).then(resolve, reject);
    },
  );

  it(
    'should support setting a @client @export variable, loaded via a ' +
      'local resolver, on a virtual field that is combined into a remote query.',
    done => {
      const query = gql`
        query postRequiringReview($reviewerId: Int!) {
          postRequiringReview {
            id
            title
            currentReviewer @client {
              id @export(as: "reviewerId")
            }
          }
          reviewerDetails(reviewerId: $reviewerId) {
            name
          }
        }
      `;

      const postRequiringReview = {
        id: 10,
        title: 'The Local State Conundrum',
        __typename: 'Post',
      };
      const reviewerDetails = {
        name: 'John Smith',
        __typename: 'Reviewer',
      };
      const currentReviewer = {
        id: 100,
        __typename: 'CurrentReviewer',
      };

      const link = new ApolloLink(({ variables }) => {
        expect(variables).toMatchObject({ reviewerId: currentReviewer.id });
        return Observable.of({
          data: {
            postRequiringReview,
            reviewerDetails,
          },
        });
      });

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {
          Post: {
            currentReviewer() {
              return currentReviewer;
            },
          },
        },
      });

      cache.writeQuery({
        query,
        data: {
          postRequiringReview: {
            __typename: 'Post',
          },
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          postRequiringReview: {
            id: postRequiringReview.id,
            title: postRequiringReview.title,
            currentReviewer,
          },
          reviewerDetails,
        });
        done();
      });
    },
  );

  it(
    'should support combining @client @export variables, calculated by a ' +
      'local resolver, with remote mutations',
    done => {
      const mutation = gql`
        mutation upvotePost($postId: Int!) {
          topPost @client @export(as: "postId")
          upvotePost(postId: $postId) {
            title
            votes
          }
        }
      `;

      const testPostId = 100;
      const testPost = {
        title: 'The Day of the Jackal',
        votes: 10,
        __typename: 'post',
      };

      const link = new ApolloLink(({ variables }) => {
        expect(variables).toMatchObject({ postId: testPostId });
        return Observable.of({
          data: {
            upvotePost: testPost,
          },
        });
      });

      const client = new ApolloClient({
        cache: new InMemoryCache(),
        link,
        resolvers: {
          Mutation: {
            topPost() {
              return testPostId;
            },
          },
        },
      });

      return client.mutate({ mutation }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          upvotePost: testPost,
        });
        done();
      });
    },
  );

  it(
    'should support combining @client @export variables, calculated by ' +
      'reading from the cache, with remote mutations',
    done => {
      const mutation = gql`
        mutation upvotePost($postId: Int!) {
          topPost @client @export(as: "postId")
          upvotePost(postId: $postId) {
            title
            votes
          }
        }
      `;

      const testPostId = 100;
      const testPost = {
        title: 'The Day of the Jackal',
        votes: 10,
        __typename: 'post',
      };

      const link = new ApolloLink(({ variables }) => {
        expect(variables).toMatchObject({ postId: testPostId });
        return Observable.of({
          data: {
            upvotePost: testPost,
          },
        });
      });

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      cache.writeQuery({
        query: gql`{ topPost }`,
        data: {
          topPost: testPostId,
        },
      });

      return client.mutate({ mutation }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          upvotePost: testPost,
        });
        done();
      });
    },
  );

  it('should not add __typename to @export-ed objects (#4691)', () => {
    const query = gql`
      query GetListItems($where: LessonFilter) {
        currentFilter @client @export(as: "where") {
          title_contains
          enabled
        }
        lessonCollection(where: $where) {
          items {
            title
            slug
          }
        }
      }
    `;

    const expectedServerQuery = gql`
      query GetListItems($where: LessonFilter) {
        lessonCollection(where: $where) {
          items {
            title
            slug
            __typename
          }
          __typename
        }
      }
    `;

    const currentFilter = {
      title_contains: 'full',
      enabled: true,
    };

    const data = {
      lessonCollection: {
        __typename: 'LessonCollection',
        items: [
          {
            __typename: 'ListItem',
            title: 'full title',
            slug: 'slug-title',
          },
        ],
      },
    };

    const client = new ApolloClient({
      link: new ApolloLink(request => {
        expect(request.variables.where).toEqual(currentFilter);
        expect(print(request.query)).toBe(print(expectedServerQuery));
        return Observable.of({ data });
      }),
      cache: new InMemoryCache({
        addTypename: true,
      }),
      resolvers: {
        Query: {
          currentFilter() {
            return currentFilter;
          },
        },
      },
    });

    return client.query({ query }).then(result => {
      expect(result.data).toEqual({
        currentFilter,
        ...data,
      });
    });
  });

  it(
    'should use the value of the last @export variable defined, if multiple ' +
      'variables are defined with the same name',
    done => {
      const query = gql`
        query reviewerPost($reviewerId: Int!) {
          primaryReviewerId @client @export(as: "reviewerId")
          secondaryReviewerId @client @export(as: "reviewerId")
          post(reviewerId: $reviewerId) {
            title
          }
        }
      `;

      const post = {
        title: 'The One Post to Rule Them All',
        __typename: 'Post',
      };
      const primaryReviewerId = 100;
      const secondaryReviewerId = 200;

      const link = new ApolloLink(({ variables }) => {
        expect(variables).toMatchObject({ reviewerId: secondaryReviewerId });
        return Observable.of({
          data: {
            post,
          },
        });
      });

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      cache.writeQuery({
        query,
        data: {
          primaryReviewerId,
          secondaryReviewerId,
        },
      });

      return client.query({ query }).then(({ data }: any) => {
        expect({ ...data }).toMatchObject({
          post,
        });
        done();
      });
    },
  );

  it(
    'should refetch if an @export variable changes, the current fetch ' +
    'policy is not cache-only, and the query includes fields that need to ' +
    'be resolved remotely',
    done => {
      const query = gql`
        query currentAuthorPostCount($authorId: Int!) {
          currentAuthorId @client @export(as: "authorId")
          postCount(authorId: $authorId)
        }
      `;

      const testAuthorId1 = 100;
      const testPostCount1 = 200;

      const testAuthorId2 = 101;
      const testPostCount2 = 201;

      let resultCount = 0;

      const link = new ApolloLink(() =>
        Observable.of({
          data: {
            postCount: resultCount === 0 ? testPostCount1 : testPostCount2
          },
        }),
      );

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      client.writeQuery({
        query,
        data: { currentAuthorId: testAuthorId1 },
      });

      const obs = client.watchQuery({ query });
      obs.subscribe({
        next({ data }) {
          if (resultCount === 0) {
            expect({ ...data }).toMatchObject({
              currentAuthorId: testAuthorId1,
              postCount: testPostCount1,
            });
            client.writeQuery({
              query,
              data: { currentAuthorId: testAuthorId2 },
            });
          } else if (resultCount === 1) {
            expect({ ...data }).toMatchObject({
              currentAuthorId: testAuthorId2,
              postCount: testPostCount2,
            });
            done();
          }
          resultCount += 1;
        }
      });
    }
  );

  it(
    'should NOT refetch if an @export variable has not changed, the ' +
    'current fetch policy is not cache-only, and the query includes fields ' +
    'that need to be resolved remotely',
    done => {
      const query = gql`
        query currentAuthorPostCount($authorId: Int!) {
          currentAuthorId @client @export(as: "authorId")
          postCount(authorId: $authorId)
        }
      `;

      const testAuthorId1 = 100;
      const testPostCount1 = 200;

      const testPostCount2 = 201;

      let resultCount = 0;

      let fetchCount = 0;
      const link = new ApolloLink(() => {
        fetchCount += 1;
        return Observable.of({
          data: {
            postCount: testPostCount1
          },
        });
      });

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      client.writeQuery({
        query,
        data: { currentAuthorId: testAuthorId1 },
      });

      const obs = client.watchQuery({ query });
      obs.subscribe({
        next(result) {
          if (resultCount === 0) {
            expect(fetchCount).toBe(1);
            expect(result.data).toMatchObject({
              currentAuthorId: testAuthorId1,
              postCount: testPostCount1,
            });

            client.writeQuery({
              query,
              variables: { authorId: testAuthorId1 },
              data: { postCount: testPostCount2 }
            });
          } else if (resultCount === 1) {
            // Should not have refetched
            expect(fetchCount).toBe(1);
            done();
          }

          resultCount +=1;
        },
      });
    }
  );

  it(
    'should NOT attempt to refetch over the network if an @export variable ' +
    'has changed, the current fetch policy is cache-first, and the remote ' +
    'part of the query (that leverages the @export variable) can be fully ' +
    'found in the cache.',
    done => {
      const query = gql`
        query currentAuthorPostCount($authorId: Int!) {
          currentAuthorId @client @export(as: "authorId")
          postCount(authorId: $authorId)
        }
      `;

      const testAuthorId1 = 1;
      const testPostCount1 = 100;

      const testAuthorId2 = 2;
      const testPostCount2 = 200;

      let fetchCount = 0;
      const link = new ApolloLink(() => {
        fetchCount += 1;
        return Observable.of({
          data: {
            postCount: testPostCount1
          },
        });
      });

      const cache = new InMemoryCache();
      const client = new ApolloClient({
        cache,
        link,
        resolvers: {},
      });

      client.writeQuery({
        query: gql`{ currentAuthorId }`,
        data: { currentAuthorId: testAuthorId1 }
      });

      let resultCount = 0;
      const obs = client.watchQuery({ query, fetchPolicy: 'cache-first' });
      obs.subscribe({
        next(result) {
          if (resultCount === 0) {
            // The initial result is fetched over the network.
            expect(fetchCount).toBe(1);
            expect(result.data).toMatchObject({
              currentAuthorId: testAuthorId1,
              postCount: testPostCount1,
            });

            client.writeQuery({
              query,
              variables: { authorId: testAuthorId2 },
              data: { postCount: testPostCount2 }
            });
            client.writeQuery({
              query: gql`{ currentAuthorId }`,
              data: { currentAuthorId: testAuthorId2 }
            });
          } else if (resultCount === 1) {
            // The updated result should not have been fetched over the
            // network, as it can be found in the cache.
            expect(fetchCount).toBe(1);
            expect(result.data).toMatchObject({
              currentAuthorId: testAuthorId2,
              postCount: testPostCount2,
            });
            done();
          }

          resultCount += 1;
        },
      });
    }
  );

  it(
    "should update @client @export variables on each broadcast if they've " +
    "changed",
    done => {
      const cache = new InMemoryCache();

      const widgetCountQuery = gql`{ widgetCount @client }`;
      cache.writeQuery({
        query: widgetCountQuery,
        data: {
          widgetCount: 100
        }
      });

      const client = new ApolloClient({
        cache,
        resolvers: {
          Query: {
            doubleWidgets(_, { widgetCount }) {
              return widgetCount ? widgetCount * 2 : 0;
            }
          }
        }
      });

      const doubleWidgetsQuery = gql`
        query DoubleWidgets($widgetCount: Int!) {
          widgetCount @client @export(as: "widgetCount")
          doubleWidgets(widgetCount: $widgetCount) @client
        }
      `;

      let count = 0;
      const obs = client.watchQuery({ query: doubleWidgetsQuery });
      obs.subscribe({
        next({ data }) {
          switch (count) {
            case 0:
              expect(data.widgetCount).toEqual(100);
              expect(data.doubleWidgets).toEqual(200);

              client.writeQuery({
                query: widgetCountQuery,
                data: {
                  widgetCount: 500
                }
              });
              break;
            case 1:
              expect(data.widgetCount).toEqual(500);
              expect(data.doubleWidgets).toEqual(1000);
              done();
              break;
            default:
          }
          count += 1;
        },
      });
    }
  );
});
