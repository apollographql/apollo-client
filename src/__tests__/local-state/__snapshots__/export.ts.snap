// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@client @export tests should NOT refetch if an @export variable has not changed, the current fetch policy is not cache-only, and the query includes fields that need to be resolved remotely 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'postCount' while writing result {
  \\"currentAuthorId\\": 100
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`@client @export tests should allow @client @export variables to be used with remote queries 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'postCount' while writing result {
  \\"currentAuthor\\": {
    \\"name\\": \\"John Smith\\",
    \\"authorId\\": 100,
    \\"__typename\\": \\"Author\\"
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`@client @export tests should refetch if an @export variable changes, the current fetch policy is not cache-only, and the query includes fields that need to be resolved remotely 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'postCount' while writing result {
  \\"appContainer\\": {
    \\"systemDetails\\": {
      \\"currentAuthor\\": {
        \\"name\\": \\"John Smith\\",
        \\"authorId\\": 100,
        \\"__typename\\": \\"Author\\"
      },
      \\"__typename\\": \\"SystemDetails\\"
    },
    \\"__typename\\": \\"AppContainer\\"
  }
}",
    ],
    Array [
      "Missing field 'title' while writing result {
  \\"loggedInReviewerId\\": 100,
  \\"__typename\\": \\"Post\\",
  \\"id\\": 10
}",
    ],
    Array [
      "Missing field 'reviewerDetails' while writing result {
  \\"postRequiringReview\\": {
    \\"loggedInReviewerId\\": 100,
    \\"__typename\\": \\"Post\\",
    \\"id\\": 10
  }
}",
    ],
    Array [
      "Missing field 'id' while writing result {
  \\"__typename\\": \\"Post\\"
}",
    ],
    Array [
      "Missing field 'title' while writing result {
  \\"__typename\\": \\"Post\\"
}",
    ],
    Array [
      "Missing field 'reviewerDetails' while writing result {
  \\"postRequiringReview\\": {
    \\"__typename\\": \\"Post\\"
  }
}",
    ],
    Array [
      "Missing field 'post' while writing result {
  \\"primaryReviewerId\\": 100,
  \\"secondaryReviewerId\\": 200
}",
    ],
    Array [
      "Missing field 'postCount' while writing result {
  \\"currentAuthorId\\": 100
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
